File: problem/Best_Time_Buy_Sell/maxProfit.js
/**
 * @param {number[]} prices
 * @param {number[]} strategy
 * @param {number} k
 * @return {number}
 */
var maxProfit = function(prices, strat, k) {
  const n = prices.length;
  const h = Math.floor(k / 2);

  const sp = new Array(n);
  let base = 0n;
  for (let i = 0; i < n; i++) {
    sp[i] = BigInt(strat[i]) * BigInt(prices[i]);
    base += sp[i];
  }

  let ans;
  if (n === k) {
    const winOrig = base;
    let winMod = 0n;
    for (let i = h; i < n; i++) winMod += BigInt(prices[i]);
    const change = winMod - winOrig;
    ans = base + (change > 0n ? change : 0n);
  } else {
    let winOrig = 0n;
    for (let i = 0; i < k; i++) winOrig += sp[i];

    let winMod = 0n;
    for (let i = h; i < k; i++) winMod += BigInt(prices[i]);

    let maxCh = winMod - winOrig;

    for (let i = 1; i <= n - k; i++) {
      winOrig += sp[i + k - 1] - sp[i - 1];
      winMod += BigInt(prices[i + k - 1]) - BigInt(prices[i - 1 + h]);
      const diff = winMod - winOrig;
      if (diff > maxCh) maxCh = diff;
    }

    ans = base + (maxCh > 0n ? maxCh : 0n);
  }

  return Number(ans);
};

File: problem/Best_Time_Buy_Sell/maxProfit.java
class Solution {
    public long maxProfit(int[] prices, int[] strat, int k) {
        int n = prices.length;
        int h = k / 2;
        long[] sp = new long[n];
        long base = 0;
        for (int i = 0; i < n; i++) {
            sp[i] = (long)strat[i] * prices[i];
            base += sp[i];
        }

        if (n == k) {
            long winOrig = base;
            long winMod = 0;
            for (int i = h; i < n; i++) winMod += prices[i];
            long change = winMod - winOrig;
            return base + Math.max(0, change);
        }

        long winOrig = 0;
        for (int i = 0; i < k; i++) winOrig += sp[i];

        long winMod = 0;
        for (int i = h; i < k; i++) winMod += prices[i];

        long maxCh = winMod - winOrig;

        for (int i = 1; i <= n - k; i++) {
            winOrig += sp[i + k - 1] - sp[i - 1];
            winMod += prices[i + k - 1] - prices[i - 1 + h];
            maxCh = Math.max(maxCh, winMod - winOrig);
        }

        return base + Math.max(0, maxCh);
    }
}